File: OrchidCore/src/main/java/com/eden/orchid/impl/themes/functions/LoadFunction.java
package com.eden.orchid.impl.themes.functions;

import com.eden.orchid.api.OrchidContext;
import com.eden.orchid.api.compilers.TemplateFunction;
import com.eden.orchid.api.converters.StringConverter;
import com.eden.orchid.api.options.annotations.BooleanDefault;
import com.eden.orchid.api.options.annotations.Description;
import com.eden.orchid.api.options.annotations.Option;
import com.eden.orchid.api.resources.resource.OrchidResource;
import lombok.Getter;
import lombok.Setter;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public final class LoadFunction extends TemplateFunction {

    private final OrchidContext context;
    private final StringConverter converter;

    @Getter @Setter
    @Option
    @Description("The resource to lookup and render in-place.")
    private String resource;

    @Getter @Setter
    @Option @BooleanDefault(true)
    @Description("When true, only resources from local sources are considered, otherwise resources from all plugins " +
            "and from the current theme will also be considered."
    )
    private boolean localResourcesOnly;

    @Inject
    public LoadFunction(OrchidContext context, StringConverter converter) {
        super("load", true);
        this.context = context;
        this.converter = converter;
    }

    @Override
    public String[] parameters() {
        return new String[] {"resource", "localResourcesOnly"};
    }

    @Override
    public Object apply(Object input) {
        String resourceName = converter.convert((input != null) ? input : resource).second;

        OrchidResource resource = localResourcesOnly ? context.getLocalResourceEntry(resourceName) : context.getResourceEntry(resourceName);

        if(resource != null) {
            return resource.getContent();
        }

        return "";
    }

}

File: OrchidCore/src/main/kotlin/com/eden/orchid/impl/themes/functions/LoadFunction.kt
package com.eden.orchid.impl.themes.functions

import com.eden.orchid.api.OrchidContext
import com.eden.orchid.api.compilers.TemplateFunction
import com.eden.orchid.api.options.annotations.BooleanDefault
import com.eden.orchid.api.options.annotations.Description
import com.eden.orchid.api.options.annotations.Option
import com.eden.orchid.api.theme.pages.OrchidPage

@Description(value = "Load content from a resource.", name = "Load")
class LoadFunction : TemplateFunction("load", false) {

    @Option
    @Description("The resource to lookup.")
    lateinit var resource: String

    @Option
    @BooleanDefault(true)
    @Description("When true, only resources from local sources are considered, otherwise resources from all plugins "
            + "and from the current theme will also be considered."
    )
    var local: Boolean = true

    @Option
    @BooleanDefault(false)
    @Description("When true, return the Front Matter data from the resource, instead of its content.")
    var frontMatter: Boolean = false

    @Option
    @BooleanDefault(true)
    @Description("When true, content will automatically be compiled according to its extension. If false, the raw " +
            "content will be returned directly as a String, minus any Front Matter."
    )
    var compile: Boolean = true

    override fun parameters(): Array<String> {
        return arrayOf("resource", "local", "compile")
    }

    override fun apply(context: OrchidContext, page: OrchidPage?): Any? {
        val foundResource = if (local) context.getLocalResourceEntry(resource) else context.getResourceEntry(resource)

        return if (foundResource != null) {
            if (frontMatter) {
                foundResource.embeddedData.element
            }
            else if (compile) {
                foundResource.compileContent(null)
            }
            else {
                foundResource.content
            }
        }
        else ""
    }
}
